### ISSUE TITLE
Gradients for the pose

### ISSUE THREAD

**Issue by pxl-th** [nianticlabs/monodepth2#400]
Hi! I have a question about gradients for the pose network.
Assume `r` and `t` are the outputs from the pose network, point `x` is then transformed by `[r|t]` and loss is the distance from point to the origin.
If the example below is correct, it seems there are no gradients for the rotational component.
If this is true, does it mean there is no contribution from the rotational part on pose-network parameter update?

```python
x = tensor([1, 2, 3, 1], dtype=float32).reshape((1, 4, 1))
r = rand((1, 1, 3), dtype=float32, requires_grad=True)
t = rand((1, 1, 3), dtype=float32, requires_grad=True)

T = transformation_from_parameters(r, t)
loss = sum(norm(matmul(T, x)[:, 0:3, :], dim=2))
loss.backward()

print(r.grad) # == None
print(t.grad) # == 1s
```

**Comment by mrharicot** [comment]
Hi,
Which pytorch version are you using?
We just did some tests and we could reproduce with `1.10` but could not with `0.4.1`, `1.7`, or `1.9`

**Comment by mrharicot** [comment]
Honestly not sure what the problem is, but here's a fixed version which works with pytorch 1.10
```python
def rot_from_axisangle(vec_b13):
    """Convert an axisangle rotation into a 4x4 transformation matrix
    (adapted from https://github.com/Wallacoloo/printipi)
    Input 'vec' has to be Bx1x3
    """
    
    vec_b3 = vec_b13.squeeze(1)
    
    angle = torch.norm(vec_b3, dim=1, keepdim=False)
    axis = vec_b3 / (angle.unsqueeze(-1) + 1e-7)

    ca = torch.cos(angle)
    sa = torch.sin(angle)
    C = 1 - ca

    x, y, z = torch.unbind(axis, -1)

    xs = x * sa
    ys = y * sa
    zs = z * sa
    xC = x * C
    yC = y * C
    zC = z * C
    xyC = x * yC
    yzC = y * zC
    zxC = z * xC

    rot = torch.zeros((r.shape[0], 4, 4))

    rot[:, 0, 0] = x * xC + ca
    rot[:, 0, 1] = xyC - zs
    rot[:, 0, 2] = zxC + ys
    rot[:, 1, 0] = xyC + zs
    rot[:, 1, 1] = y * yC + ca
    rot[:, 1, 2] = yzC - xs
    rot[:, 2, 0] = zxC - ys
    rot[:, 2, 1] = yzC + xs
    rot[:, 2, 2] = z * zC + ca
    rot[:, 3, 3] = 1

    return rot
```

**Comment by mrharicot** [comment]
Thanks to @mdfirman we found a minimal example to reproduce the bug.
I will file an issue on the pytorch github.

```python
import torch

r = torch.rand((1), dtype=torch.float32, requires_grad=True)
rot = torch.zeros((1, 1))

rot[:, 0] = torch.squeeze(r) #this breaks in 1.10 but works fine before
rot[:, 0] = r                #this works in all versions

loss = torch.mean(rot)
loss.backward()

print(r.grad)
```

**Comment by pxl-th** [comment]
Yes, only PyTorch 1.10 seems to be affected. Thanks for posting version that works, I meanwhile resorted to using [so3_exp_map](https://github.com/facebookresearch/pytorch3d/blob/main/pytorch3d/transforms/so3.py#L148) from pytorch3d.
